cocos/renderer/gfx-validator/BufferValidator.h

```diff
@@ -0,0 +1,51 @@
+/****************************************************************************
+ Copyright (c) 2020-2021 Xiamen Yaji Software Co., Ltd.
+
+ http://www.cocos.com
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+****************************************************************************/
+
+#pragma once
+
+#include "base/Agent.h"
+#include "gfx-base/GFXBuffer.h"
+
+namespace cc {
+namespace gfx {
+
+class CC_DLL BufferValidator final : public Agent<Buffer> {
+public:
+    using Agent::Agent;
+    ~BufferValidator() override;
+
+    void update(void *buffer, uint size) override;
+
+protected:
+    void doInit(const BufferInfo &info) override;
+    void doInit(const BufferViewInfo &info) override;
+    void doResize(uint size, uint count) override;
+    void doDestroy() override;
+
+    vector<uint8_t> _buffer;
```

Comment:
Should record it here? Can use actor's buffer? Or it will consume too much memory.

Reply:
The validator doesn't know about the underlying backend's data structure. It has to maintain its own data.

Reply:
Yep, you are right. But does it need to record data? Will the validation layer check its content?

Reply:
Most likely not. that's why the actual allocation only happens inside 'if (device->isRecording())' conditions

Reply:
Sounds good.